substra/sdk/backends/local/backend.py

```
@@ -0,0 +1,493 @@
+                if compute_plan.aggregatetuple_keys is not None
+                else 0
+            )
+            #  Add to the compute plan
+            if compute_plan.traintuple_keys is None:
+                compute_plan.traintuple_keys = [key]
+            else:
+                compute_plan.traintuple_keys.append(key)
+            compute_plan.tuple_count += 1
+            compute_plan.status = models.Status.waiting
+
+        else:
+            compute_plan_id = ""
+            rank = 0
+
+        # create model
+        options = {}
+        traintuple = models.Traintuple(
+            key=key,
+            creator=_BACKEND_ID,
+            worker=_BACKEND_ID,
+            algo_key=spec.algo_key,
+            dataset={
+                "opener_hash": spec.data_manager_key,
+                "keys": spec.train_data_sample_keys,
+                "worker": _BACKEND_ID,
+            },
+            permissions={
+                "process": {"public": public, "authorized_ids": authorized_ids},
+            },
+            log="",
+            compute_plan_id=compute_plan_id,
+            rank=rank,
+            tag=spec.tag or "",
+            status=models.Status.waiting.value,
+            in_models=[
+                {
+                    "hash": in_traintuple.out_model.hash_,
+                    "storage_address": in_traintuple.out_model.storage_address,
+                }
+                for in_traintuple in in_traintuples
+            ],
+            **options,
+        )
+
+        traintuple = self._db.add(traintuple, exist_ok)
+        self._worker.schedule_traintuple(traintuple)
+        return traintuple
+
+    def _add_testtuple(self, spec, exist_ok, spec_options=None):
+
+        # validation
+        objective = self._db.get(schemas.Type.Objective, spec.objective_key)
+        traintuple = self._db.get(schemas.Type.Traintuple, spec.traintuple_key)
+        assert traintuple.out_model is not None
```

I think we should raise a proper exception here for the user.